src/H5ES.c

```diff
@@ -235,6 +235,61 @@ H5ESget_op_counter(hid_t es_id, uint64_t *op_counter /*out*/)
     FUNC_LEAVE_API(ret_value)
 } /* end H5ESget_op_counter() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5ESget_requests
+ *
+ * Purpose:     Retrieve the requests in an event set.  Up to *count
+ *              requests are stored in the provided requests array, and
+ *              the connector ids corresponding to these requests are
+ *              stored in the provided connector_ids array.  Either or
+ *              both of these arrays may be NULL, in which case this
+ *              information is not returned.  If these arrays are
+ *              non-NULL, they must be large enough to contain *count
+ *              entries.  On exit, *count is set to the total number of
+ *              events in the event set.
+ *
+ *              Events are returned in the order they were added to the
+ *              event set.  If order is H5_ITER_INC or H5_ITER_NATIVE,
+ *              events will be returned starting from the oldest. If order
+ *              is H5_ITER_DEC, events will be returned starting with the
+ *              newest/most recent.
+ *
+ * Return:      SUCCEED / FAIL
+ *
+ * Programmer:  Neil Fortner
+ *              Tuesday, November 23, 2021
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5ESget_requests(hid_t es_id, H5_iter_order_t order, hid_t *connector_ids, void **requests,
+                 size_t *count /*out*/)
+{
+    H5ES_t *es;                  /* Event set */
+    herr_t  ret_value = SUCCEED; /* Return value */
+
+    FUNC_ENTER_API(FAIL)
+    H5TRACE5("e", "iIo*i**xx", es_id, order, connector_ids, requests, count);
+
+    /* Check arguments */
+    if (NULL == (es = H5I_object_verify(es_id, H5I_EVENTSET)))
+        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "invalid event set identifier")
+    if (order <= H5_ITER_UNKNOWN || order >= H5_ITER_N)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid iteration order specified")
+
+    /* Call internal routine */
+    if (count && *count > 0 && (requests || connector_ids))
+        if (H5ES__get_requests(es, order, connector_ids, requests, *count) < 0)
+            HGOTO_ERROR(H5E_EVENTSET, H5E_CANTGET, FAIL, "can't get requests")
+
+    /* Retrieve the count, if non-NULL */
+    if (count)
+        *count = H5ES__list_count(&es->active);
+
```

Comment:
Not sure if you should allow for count to be NULL if the routine does nothing if count is NULL :)

Reply:
Well free() will happily do nothing if you pass NULL ;)  Though I don't have a strong feeling one way or the other.

In general though, I should mention that I thought this coding pattern (passing a pointer to the length of a buffer that is updated during the routine) was common throughout the library, but now I can't find any examples of it in the public API. Perhaps it is only used for internal interfaces. I'm open to changing the interface, though having the function return both the data and the length of the data in a single call may prevent some race conditions in a multithreaded application (with threadsafe on). Certainly we should always pass the length of the user's buffer.

Reply:
API revised according to discussion on Friday. There are now separate input and output parameters array_len and count. The function now does something even if array_len is 0 (it returns count). I don't know if it's worth throwing an error if array_len is 0 and count is NULL.